### PR TITLE
fix: launcher does not kill firefox.exe on WSL

### PR DESCRIPTION
**Pull request by dotnetCarpenter**
**Record PIDs for firefox.exe on WSL and kill all processes not recorded on start-up.**

After testing both regular Firefox and Firefox Developer Edition, I came to the conclusion that the executable is always `firefox.exe` - I really hope that is correct.

At first I wrote this patch in ES5 but since we already use destructuring assignment at the top (`var { execSync } = require('child_process')`), I re-wrote it to ES6/ES2015.

According to SO the highest PID on Windows is [0xFFFFFFFC](https://stackoverflow.com/questions/17868218/what-is-the-maximum-process-id-on-windows) which I believe is 10 digits. The lowest PID is 0 but only PIDs above 100 seems to be user-land processes. The first ones seems to be reserved for Windows OS. The way I extract PIDs is by looking for _whole_ integers, so I could look for 2 digits PIDs but never 1, since we would then get Session Number as PID. If a PID in the 0-9 range can be a Firefox process, then `extractPids` logic needs to change. But I very much doubt that. Currently I test for integers with 3-10 digits.

This PR passes all tests on my machine on WSL, although it will kill the second test mid-air but then Karma retries and everything works out. [Karma-browserstack-launcher uses a different exit/kill event](https://github.com/karma-runner/karma-browserstack-launcher/blob/master/index.js#L29) but that looks specific to their launcher and might not help at all.

Ref: https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143

### CODE REVIEW

File: index.js
Comment (birtles) [inline index.js:50]: Should this be `log.error` ?

File: index.js
Comment (birtles) [inline index.js:298]: Nit: It seems odd to me to execute the code to kill the tasks as an argument to `log.debug()`. I think many people will read the code and skip over the `log.debug()` line assuming it is just for logging.

Can we separate this out into two lines? One that kills the tasks and records the result, and one that logs it?

**Review by birtles — APPROVED** [review]
This looks great to me. I left a few nits but feel free to push back on them.

File: index.js
Comment (birtles) [inline index.js:297]: I guess this should be `log.info` too. It seems a bit odd to display warning when everything is operating as expected.

**Comment by birtles** [comment]
> After testing both regular Firefox and Firefox Developer Edition, I came to the conclusion that the executable is always `firefox.exe` - I really hope that is correct.

I think it should be fine. If we encounter cases where that's not true, we can refine it then.

> At first I wrote this patch in ES5 but since we already use destructuring assignment at the top (`var { execSync } = require('child_process')`), I re-wrote it to ES6/ES2015.

Yes, sounds good.

> According to SO the highest PID on Windows is [0xFFFFFFFC](https://stackoverflow.com/questions/17868218/what-is-the-maximum-process-id-on-windows) which I believe is 10 digits. The lowest PID is 0 but only PIDs above 100 seems to be user-land processes. The first ones seems to be reserved for Windows OS. The way I extract PIDs is by looking for _whole_ integers, so I could look for 2 digits PIDs but never 1, since we would then get Session Number as PID. If a PID in the 0-9 range can be a Firefox process, then `extractPids` logic needs to change. But I very much doubt that. Currently I test for integers with 3-10 digits.

Yes, that makes sense. Thanks for being so thorough.

> This PR passes all tests on my machine on WSL, although it will kill the second test mid-air but then Karma retries and everything works out. [Karma-browserstack-launcher uses a different exit/kill event](https://github.com/karma-runner/karma-browserstack-launcher/blob/master/index.js#L29) but that looks specific to their launcher and might not help at all.

I just tried this locally. The first time I ran into issues connecting to localhost because WSL seems to be flaky like that. After restarting WSL, however, I got further.

I get a `Gah, your tab just crashed` error in Firefox.

In the console I get:

```log
...
07 08 2021 11:00:57.637:INFO [karma-server]: Karma v4.2.0 server started at http://0.0.0.0:9876/
07 08 2021 11:00:57.639:INFO [launcher]: Launching browsers Firefox, FirefoxHeadless with concurrency unlimited
07 08 2021 11:00:57.644:INFO [launcher]: Starting browser Firefox
07 08 2021 11:00:57.657:WARN [FirefoxLauncher]: WSL environment detected: Please do not open Firefox while running tests as it will be killed after the test!
07 08 2021 11:00:57.658:WARN [FirefoxLauncher]: WSL environment detected: See https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143
07 08 2021 11:00:57.963:INFO [launcher]: Starting browser FirefoxHeadless
07 08 2021 11:00:57.977:WARN [FirefoxLauncher]: WSL environment detected: Please do not open Firefox while running tests as it will be killed after the test!
07 08 2021 11:00:57.977:WARN [FirefoxLauncher]: WSL environment detected: See https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143
07 08 2021 11:01:05.765:INFO [Firefox 90.0.0 (Windows 10.0.0)]: Connected on socket Z1N_LGlRSpe_-TrQAAAA with id 59932439
.
Firefox 90.0.0 (Windows 10.0.0): Executed 1 of 1 SUCCESS (0.003 secs / 0 secs)
07 08 2021 11:01:06.215:WARN [FirefoxLauncher]: Killing the following PIDs: [
  '2696',  '1788',
  '28836', '24164',
  '12664', '30600',
  '31944', '22560',
  '3248',  '20676',
  '13016', '24816'
]
ERROR: The process "30600" not found.
07 08 2021 11:01:06.390:WARN [FirefoxLauncher]: Error: Command failed: taskkill.exe /F /PID 2696 /PID 1788 /PID 28836 /PID 24164 /PID 12664 /PID 30600 /PID 31944 /PID 22560 /PID 3248 /PID 20676 /PID 13016 /PID 24816
ERROR: The process "30600" not found.
```

If I manually close the Firefox window it restarts and succeeds.

I guess the `FirefoxHeadless` launcher ends up killing the content process for the `Firefox` launcher?

If we could pass different names to `logger.create()` for each of the different launchers I guess that would make debugging this easier.

The second time I tried around I got a similar situation but with a slightly different log:

```log
...
07 08 2021 10:55:54.925:INFO [Firefox 90.0.0 (Windows 10.0.0)]: Connected on socket J5Jh_hSsZ4fhfnYdAAAA with id 98877858
.
Firefox 90.0.0 (Windows 10.0.0): Executed 1 of 1 SUCCESS (0.002 secs / 0 secs)
07 08 2021 10:55:55.383:WARN [FirefoxLauncher]: Killing the following PIDs: [
  '26992', '24120',
  '26408', '27508',
  '14524', '13724',
  '13552', '13192',
  '32444', '13400',
  '7376',  '17016'
]
ERROR: The process with PID 13552 could not be terminated.
Reason: There is no running instance of the task.
```

Again, manually closing the Firefox window lets it restart and the second time around it succeeds.

File: index.js
Comment (dotnetCarpenter) [reply index.js:50]: My thinking was that when multiple Firefox browsers are used then `taskkill.exe` might throw without it being an issue when multiple `taskkill.exe` commands are issued for PIDs that are already killed. But I just tested that and `taskkill.exe` does not throw in that case but simply returns:

```
ERROR: The process with PID 11464 could not be terminated.
Reason: There is no running instance of the task.
```

So I will change it to `log.error` :)

File: index.js
Comment (dotnetCarpenter) [reply index.js:298]: Sure

File: index.js
Comment (dotnetCarpenter) [reply index.js:50]: Looking at your first log, it looks like you triggered an throwable error in `safeExecSync` without it being a _"real"_ error.

> 07 08 2021 11:01:06.390:WARN [FirefoxLauncher]: Error: Command failed: taskkill.exe /F /PID 2696 /PID 1788 /PID 28836 /PID 24164 /PID 12664 /PID 30600 /PID 31944 /PID 22560 /PID 3248 /PID 20676 /PID 13016 /PID 24816

In that case, I think it should only be a warning since the test will run as usual.

**Comment by dotnetCarpenter** [comment]
I am seeing the same as you when running with multiple Firefox browsers.

>If I manually close the Firefox window it restarts and succeeds.

You do not have to manually close Firefox if you are patient. Eventually the browser will get killed and restarted. `:WARN [launcher]: Firefox have not captured in 60000 ms, killing.`

But I can help think that `this.on ('kill', function (done) {` event is fired prematurely.

With the log level set to debug we get the test running and then immediately we close the browser tab process.

```
07 08 2021 13:25:28.698:DEBUG [Firefox 90.0.0 (Windows 10.0.0)]: EXECUTING -> CONNECTED

Firefox 90.0.0 (Windows 10.0.0): Executed 1 of 1 SUCCESS (0.009 secs / 0 secs)
07 08 2021 13:25:29.741:WARN [FirefoxLauncher]: Killing the following PIDs: [
  '2860',  '11044',
  '15812', '16200',
  '9484',  '16328',
  '9136',  '3576',
  '5972',  '2996',
  '488'
]
07 08 2021 13:25:32.192:DEBUG [FirefoxLauncher]: SUCCESS: The process with PID 2860 has been terminated.
SUCCESS: The process with PID 11044 has been terminated.
SUCCESS: The process with PID 15812 has been terminated.
SUCCESS: The process with PID 16200 has been terminated.
SUCCESS: The process with PID 9484 has been terminated.
SUCCESS: The process with PID 16328 has been terminated.
SUCCESS: The process with PID 9136 has been terminated.
SUCCESS: The process with PID 3576 has been terminated.
SUCCESS: The process with PID 5972 has been terminated.
SUCCESS: The process with PID 2996 has been terminated.
SUCCESS: The process with PID 488 has been terminated.

07 08 2021 13:25:32.192:DEBUG [launcher]: CAPTURED -> BEING_KILLED
07 08 2021 13:25:32.192:DEBUG [launcher]: BEING_KILLED -> BEING_FORCE_KILLED
07 08 2021 13:25:32.194:WARN [launcher]: FirefoxHeadless was not killed in 2000 ms, sending SIGKILL.
07 08 2021 13:25:32.197:DEBUG [launcher]: Process FirefoxHeadless exited with code null and signal SIGTERM
```

I feel that `CAPTURED -> BEING_KILLED` and `BEING_KILLED -> BEING_FORCE_KILLED` should come **before** actually killing processes.

```
07 08 2021 13:25:32.192:DEBUG [launcher]: CAPTURED -> BEING_KILLED
07 08 2021 13:25:32.192:DEBUG [launcher]: BEING_KILLED -> BEING_FORCE_KILLED
```

Perhaps the *killing* should be in a timeout?

File: index.js
Comment (dotnetCarpenter) [reply index.js]: I've changed it to `log.debug`... Since it's perfectly fine.. Probably.. Might be useful to see while debugging

**Comment by dotnetCarpenter** [comment]
We spend a lot of time re-starting Firefox when running multiple Firefox browsers inadvertently kills the slowest browser. If I could somehow halt the _killing_ until the last browser finish, I think we would see a major speedup and it will be less confusing for the user running the tests. A restart adds 1 minute to the tests..

Here is the logs for first testing with Firefox and FirefoxHeadless (Done in 74.02s.), then only with Firefox (Done in 13.73s.) and lastly only with FirefoxHeadless (Done in 11.11s.):

```
dotnet@DESKTOP-NDR7PBQ:~/projects/opensource/karma-firefox-launcher (feature/kill_on_wsl)$ yarn test
yarn run v1.22.5
$ karma start --single-run
07 08 2021 14:13:10.443:INFO [karma-server]: Karma v4.2.0 server started at http://0.0.0.0:9876/
07 08 2021 14:13:10.445:INFO [launcher]: Launching browsers Firefox, FirefoxHeadless with concurrency unlimited
07 08 2021 14:13:10.450:INFO [launcher]: Starting browser Firefox
07 08 2021 14:13:10.457:WARN [FirefoxLauncher]: WSL environment detected: Please do not open Firefox while running tests as it will be killed after the test!
07 08 2021 14:13:10.457:WARN [FirefoxLauncher]: WSL environment detected: See https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143
07 08 2021 14:13:10.652:INFO [launcher]: Starting browser FirefoxHeadless
07 08 2021 14:13:10.659:WARN [FirefoxHeadlessLauncher]: WSL environment detected: Please do not open Firefox while running tests as it will be killed after the test!
07 08 2021 14:13:10.661:WARN [FirefoxHeadlessLauncher]: WSL environment detected: See https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143
07 08 2021 14:13:19.492:INFO [Firefox 90.0.0 (Windows 10.0.0)]: Connected on socket JPu9Db1RgR46LGoBAAAA with id 64498128
.
Firefox 90.0.0 (Windows 10.0.0): Executed 1 of 1 SUCCESS (0.035 secs / 0 secs)
07 08 2021 14:14:10.463:WARN [launcher]: Firefox have not captured in 60000 ms, killing.
07 08 2021 14:14:10.946:INFO [launcher]: Trying to start Firefox again (1/2).
07 08 2021 14:14:10.957:WARN [FirefoxLauncher]: WSL environment detected: Please do not open Firefox while running tests as it will be killed after the test!
07 08 2021 14:14:10.957:WARN [FirefoxLauncher]: WSL environment detected: See https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143
07 08 2021 14:14:21.211:INFO [Firefox 90.0.0 (Windows 10.0.0)]: Connected on socket BYaSbN13nXmNADkBAAAB with id 57344761
.
Firefox 90.0.0 (Windows 10.0.0): Executed 1 of 1 SUCCESS (0.002 secs / 0 secs)
TOTAL: 2 SUCCESS
Done in 74.02s.
dotnet@DESKTOP-NDR7PBQ:~/projects/opensource/karma-firefox-launcher (feature/kill_on_wsl)$ yarn test
yarn run v1.22.5
$ karma start --single-run
07 08 2021 14:26:58.773:INFO [karma-server]: Karma v4.2.0 server started at http://0.0.0.0:9876/
07 08 2021 14:26:58.775:INFO [launcher]: Launching browsers Firefox with concurrency unlimited
07 08 2021 14:26:58.781:INFO [launcher]: Starting browser Firefox
07 08 2021 14:26:58.790:WARN [FirefoxLauncher]: WSL environment detected: Please do not open Firefox while running tests as it will be killed after the test!
07 08 2021 14:26:58.790:WARN [FirefoxLauncher]: WSL environment detected: See https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143
07 08 2021 14:27:09.046:INFO [Firefox 90.0.0 (Windows 10.0.0)]: Connected on socket i47XHgKluZiGfHsoAAAA with id 13093631
.
Firefox 90.0.0 (Windows 10.0.0): Executed 1 of 1 SUCCESS (0.041 secs / 0.001 secs)
Done in 13.73s.
dotnet@DESKTOP-NDR7PBQ:~/projects/opensource/karma-firefox-launcher (feature/kill_on_wsl)$ yarn test
yarn run v1.22.5
$ karma start --single-run
07 08 2021 14:27:31.758:INFO [karma-server]: Karma v4.2.0 server started at http://0.0.0.0:9876/
07 08 2021 14:27:31.761:INFO [launcher]: Launching browsers FirefoxHeadless with concurrency unlimited
07 08 2021 14:27:31.766:INFO [launcher]: Starting browser FirefoxHeadless
07 08 2021 14:27:31.773:WARN [FirefoxHeadlessLauncher]: WSL environment detected: Please do not open Firefox while running tests as it will be killed after the test!
07 08 2021 14:27:31.773:WARN [FirefoxHeadlessLauncher]: WSL environment detected: See https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143
07 08 2021 14:27:40.032:INFO [Firefox 90.0.0 (Windows 10.0.0)]: Connected on socket ciBjrkEFdg4Z5yNCAAAA with id 82762367
.
Firefox 90.0.0 (Windows 10.0.0): Executed 1 of 1 SUCCESS (0.005 secs / 0 secs)
Done in 11.11s.
```

**Comment by dotnetCarpenter** [comment]
I found the `emitter` dependency which is also used by [karma-sauce-launcher](https://github.com/karma-runner/karma-sauce-launcher/). Using `emitter.on ('exit'` I can run code after all of the tests in all browsers has finished. That means we do not halt tests by killing slower Firefox processes.

It also makes all `kill` commands after the first, redundant. But I did not change it, just in case the first `kill` command happens before a another Firefox process is started. I haven't tested with `concurrency` setting other than `unlimited` and 1 though.

We still keep a record of Firefox processes before any test is run, so you can still have Firefox open without risking it getting killed. As long as Firefox does not spawn new processes while the tests are running.

I think this is as good as it gets with this approach.

**Comment by dotnetCarpenter** [comment]
Here is the logs for testing with Firefox and FirefoxHeadless (Done in 20.31s.)

```
yarn test
yarn run v1.22.5
$ karma start --single-run
07 08 2021 19:27:02.975:INFO [karma-server]: Karma v4.2.0 server started at http://0.0.0.0:9877/
07 08 2021 19:27:02.990:INFO [launcher]: Launching browsers Firefox, FirefoxHeadless with concurrency unlimited
07 08 2021 19:27:03.000:INFO [launcher]: Starting browser Firefox
07 08 2021 19:27:03.013:WARN [FirefoxLauncher]: WSL environment detected: Please do not open Firefox while running tests as it will be killed after the test!
07 08 2021 19:27:03.014:WARN [FirefoxLauncher]: WSL environment detected: See https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143
07 08 2021 19:27:03.422:INFO [launcher]: Starting browser FirefoxHeadless
07 08 2021 19:27:03.435:WARN [FirefoxHeadlessLauncher]: WSL environment detected: Please do not open Firefox while running tests as it will be killed after the test!
07 08 2021 19:27:03.435:WARN [FirefoxHeadlessLauncher]: WSL environment detected: See https://github.com/karma-runner/karma-firefox-launcher/issues/101#issuecomment-891850143
07 08 2021 19:27:14.370:INFO [Firefox 90.0.0 (Windows 10.0.0)]: Connected on socket plBpDEMLC80yKAgcAAAA with id 4585927
.
Firefox 90.0.0 (Windows 10.0.0): Executed 1 of 1 SUCCESS (0.035 secs / 0 secs)
07 08 2021 19:27:18.253:INFO [Firefox 90.0.0 (Windows 10.0.0)]: Connected on socket T23WGQCjCGpZ1VhYAAAB with id 92583815
.
Firefox 90.0.0 (Windows 10.0.0): Executed 1 of 1 SUCCESS (0.032 secs / 0 secs)
TOTAL: 2 SUCCESS
07 08 2021 19:27:21.002:WARN [launcher]: Firefox was not killed in 2000 ms, sending SIGKILL.
Done in 20.31s.
```

**Comment by birtles** [comment]
That looks great to me! Thank you!
I just tried it here and it works perfectly.

**Review by birtles — APPROVED** [review]

**Comment by dotnetCarpenter** [comment]
@birtles Thanks for merging :) How far off are we on a new release?

**Comment by birtles** [comment]
> @birtles Thanks for merging :) How far off are we on a new release?

Thank you!

We're just blocked on #149.

**Comment by dotnetCarpenter** [comment]
So you do not have administrative rights over karma-firefox-launcher? If you do, then you should be able to both switch the default branch to `main` and enable karma-firefox-launcher on your own Travis account.

Given the latest code changes, it might make sense to run the test on [CircleCI](https://circleci.com/blog/circleci-matrix-jobs/) where test can be run on Windows, MacOS and Linux. Last time I checked, Travis only supports Linux.

Again, if you have rights to this repo, then you should be able to login with github and set up karma-firefox-launcher.

**Comment by birtles** [comment]
> So you do not have administrative rights over karma-firefox-launcher?

Unfortunately no. I seem to have write access but that's all. Running tests cross platform sounds like a good idea -- I think Travis supports that but I don't know if we've enabled that or not in this repo.

**Comment by dotnetCarpenter** [comment]
Usually you enable that in a configuration file and not on your Travis account. For Travis the docs are here: https://config.travis-ci.com/ and you are right, [Travis supports `linux`, `osx` and `windows`](https://config.travis-ci.com/ref/os).

The link to CircleCI's documentation is in my previous https://github.com/karma-runner/karma-firefox-launcher/pull/141#issuecomment-895594100.

As the only, current, active maintainer, I think you should ask to become Admin on this repo. That will let you change the default branch to `main`.

For Travis, you could try to create your own account at https://www.travis-ci.com/ and see if you can associate a Travis set up with karma-firefox-launcher. You can then update the badge in the README, to point to your test runs on Travis. I do not think you have to be Admin to do that, but I am not sure.

**Comment by dotnetCarpenter** [comment]
Here is a link to github permission levels, https://docs.github.com/en/organizations/managing-access-to-your-organizations-repositories/repository-permission-levels-for-an-organization.

**Edit the repository's default branch** requires an Admin role.

**Comment by birtles** [comment]
It looks like Travis wants to install a GitHub app in order to work so I don't think I can add it to this repo. (I think someone with sufficient privileges just needs to login to travis-ci.com and trigger it and it should work.) @johnjbarton has helped with all the setup in the past so let's give him a chance to follow up.

**Comment by dotnetCarpenter** [comment]
@birtles Any updates? :)

**Comment by birtles** [comment]
Yes, you can see the latest here: https://github.com/karma-runner/karma-firefox-launcher/issues/149#issuecomment-897261817

**Pull request merged**
